### ISSUE TITLE
Vanilla Data Retention

### ISSUE THREAD

**Issue by Lyonlancer5** [Lyonlancer5/Unmapi#2]
[Migrated, edited from Gist]
### Rationale

The prospect of retaining all item/block info even through vanilla loading is made possible through the storage of additional NBT data, which is only normally accessible through editors, utility mods and (as of MC 1.8) commands. 

Metadata (otherwise known as "damage values") can also be used, but only in certain cases, most importantly those without "sub-blocks" or items that **do not** have states or durability based on them. NBT (Named Binary Tags) may be a better option due to its way of being retained in vanilla.

The rationale behind re-implementing how mods add stuff into the game is due to Ak12418's "world sanitation" protocol, which relies heavily on opening modded worlds on Vanilla in order to reset almost everything that mods add. This can also help eliminate cross-world entity problems (missing inventory items, misaligned item/block ID names) since the mods only touch what is already available in vanilla.

This cross-world problems can be recreated through the use of entity copiers, such as CustomNPCs' `Mob Cloner`, then spawning the same entity to another world. Since Forge assigns different IDs for the same items/blocks in different worlds, inventory contents will be changed.

When using metadata and/or NBT, it remains the same throughout different worlds, since the vanilla mechanics of the game allow for custom metadata/NBT to be loaded and stored. The items/blocks are not affected and will not disappear when loaded in vanilla (except in edge-cases where future vanilla releases might have changed how some items and blocks work) because it is saved in vanilla's own data structure.

However, the problem lies with the releases of Minecraft which have `Data Fixers` enabled, which remove invalid metadata when detected, leaving only NBT usage feasible on vanilla retention.

In short, mods' data are still there, just non-functional.
The mods will be the one to handle the metadata, while vanilla just quietly saves and loads what it sees.

**Comment by Lyonlancer5** [comment]
[Migrated, edited from Gist]
# Implementation

The requirement of this implementation is to override Forge's method of adding content into the game, most especially on the block and item side of the game. [This is a messy proof-of-concept of the overrides](https://github.com/Lyonlancer5/Unmapi/blob/1ae3b73fe9a420f13962936a491daf3c8415bc56/src/main/java/net/lyonlancer5/mcmp/unmapi/util/reflect/ReflectedRegistry.java) which uses reflection to invoke Forge internal methods and bypass some of the restrictions the Forge registry system has.

### NBT
With the flexibility of NBT tags, this is the most viable option of storing mod data into vanilla's own save data. Mods can just use events in order to replace items/blocks with their intended operations. This, however, may incur a performance penalty due to the nature of events processing.

Class overriding can be done, but since this relies heavily on modifying NBT processing instead, it may be better off using events with custom NBT handling instead :3

### Blocks without specific metadata
Class overriding can be done just to add the handling of metadata by overriding `getSubBlocks()` and its related functions. Many blocks can be done through this way in order to prevent missing blocks from occurring when loading in from vanilla, especially those with abundance of __ores__, by simply overriding _any_ of the available ores and adding metadata into them for them to do specific stuff.

Only caveat is that opening this in a future vanilla release _may_ make some blocks appear weird due to the presence of a previously-unimplemented damage value, as well as making it harder for certain blocks to do certain things. This may also break some mods which mess with interactions on blocks.

### Blocks _with_ specific metadata
This can also be done the same way as those without damage values, however, modifying _how_ the block works with both the existing and the new metadata may be time-consuming.

### Blocks with complex NBT structures
Appending to and modifying existing NBT data structures is easy, but preventing unexpected things from happening is very important. Also note the limitations of the NBT data structure and only how much can be stored.

### Items without specific metadata
Can be considered the same as blocks without specific metadata, except interaction will need to be more specific and edited in order to prevent "cross-talk" of each item's functions.

### Items with specific metadata
**TODO**

### Items with durability
**TODO**

### Items with complex NBT structures
**TODO**: highly conflicts with NBT general header.
